components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/DependencyInstallerImpl.java

```diff
@@ -0,0 +1,265 @@
+
+	public DependencyInstallerImpl(Map<String, ExtensionConfig> extensionConfigs) {
+		this.extensionConfigs = extensionConfigs;
+	}
+
+	@Override
+	public Map<String, Object> installDependenciesFor(String extensionId) throws ExtensionsInstallerException {
+		List<DependencyConfig> completedDependencies = new ArrayList<>();
+		List<DependencyConfig> failedDependencies = new ArrayList<>();
+		List<DependencyConfig> manuallyInstallableDependencies = new ArrayList<>();
+
+		ExtensionConfig extension = ExtensionsInstallerUtils.findExtension(extensionId, extensionConfigs);
+		List<DependencyConfig> dependencies = extension.getDependencies();
+		if (dependencies != null) {
+			for (DependencyConfig dependency : dependencies) {
+				if (dependency.isAutoDownloadable()) {
+					try {
+						installUsagesFor(dependency);
+						completedDependencies.add(dependency);
+					} catch (ExtensionsInstallerException e) {
+						failedDependencies.add(dependency);
+						LOGGER.error(String.format("Failed to install dependency: %s for extension: %s.",
+								dependency.getRepresentableName(), extensionId), e);
+					}
+				} else {
+					manuallyInstallableDependencies.add(dependency);
+				}
+			}
+			ExtensionInstallationStatus status = ExtensionsInstallerUtils.getInstallationStatus(
+					completedDependencies.size(), extension.getDependencies().size());
+			return ResponseEntityCreator.createExtensionInstallationResponse(
+					status, completedDependencies, failedDependencies, manuallyInstallableDependencies);
+		}
+		throw new ExtensionsInstallerException(String.format(
+				"No dependencies were specified in extension: %s", extensionId));
+	}
+
+	private void installUsagesFor(DependencyConfig dependency) throws ExtensionsInstallerException {
+		DependencyDownloadConfig download = dependency.getDownload();
+		if (download != null) {
+			try {
+				URL downloadUrl = new URL(download.getUrl());
+				String fileName = FilenameUtils.getName(downloadUrl.getPath());
+				List<File> usageDestinations = generateUsageDestinations(dependency, fileName);
+				downloadOrCopyFilesTo(usageDestinations, downloadUrl);
+			} catch (MalformedURLException e) {
+				throw new ExtensionsInstallerException(
+						String.format("Invalid download URL: %s.", download.getUrl()), e);
+			}
+		} else {
+			throw new ExtensionsInstallerException("Unable to find property: 'download'.");
+		}
+	}
+
+	private List<File> generateUsageDestinations(DependencyConfig dependency, String fileName)
+			throws ExtensionsInstallerException {
+		Set<UsageConfig> usages = dependency.getUsages();
+		if (usages != null) {
+			List<File> fileDestinations = new ArrayList<>(usages.size());
+			for (UsageConfig usage : usages) {
+				fileDestinations.add(new File(ExtensionsInstallerUtils.getDirectoryPathFor(usage), fileName));
+			}
+			return fileDestinations;
+		}
+		throw new ExtensionsInstallerException("Unable to find property: 'usages'.");
```

Comment:
isnt this thrown every time

Reply:
No.
When `usage` is not null, it will `return` within the if block.
The exception is thrown only when `usage` is not present.